src/nvim/plines.c

```diff
@@ -102,127 +75,80 @@ int linetabsize(win_T *wp, linenr_T lnum)
   return win_linetabsize(wp, lnum, ml_get_buf(wp->w_buffer, lnum), (colnr_T)MAXCOL);
 }
 
-void win_linetabsize_cts(chartabsize_T *cts, colnr_T len)
-{
-  for (; *cts->cts_ptr != NUL && (len == MAXCOL || cts->cts_ptr < cts->cts_line + len);
-       MB_PTR_ADV(cts->cts_ptr)) {
-    cts->cts_vcol += win_lbr_chartabsize(cts, NULL);
-  }
-  // check for inline virtual text after the end of the line
-  if (len == MAXCOL && cts->cts_has_virt_text && *cts->cts_ptr == NUL) {
-    win_lbr_chartabsize(cts, NULL);
-    cts->cts_vcol += cts->cts_cur_text_width_left + cts->cts_cur_text_width_right;
-  }
-}
-
 /// Prepare the structure passed to chartabsize functions.
 ///
 /// "line" is the start of the line, "ptr" is the first relevant character.
 /// When "lnum" is zero do not use inline virtual text.
-void init_chartabsize_arg(chartabsize_T *cts, win_T *wp, linenr_T lnum, colnr_T col, char *line,
-                          char *ptr)
+CSType init_charsize_arg(CharsizeArg *cts, win_T *wp, linenr_T lnum, char *line)
 {
-  cts->cts_win = wp;
-  cts->cts_vcol = col;
-  cts->cts_line = line;
-  cts->cts_ptr = ptr;
-  cts->cts_max_head_vcol = 0;
-  cts->cts_cur_text_width_left = 0;
-  cts->cts_cur_text_width_right = 0;
-  cts->cts_has_virt_text = false;
-  cts->cts_row = lnum - 1;
-
-  if (cts->cts_row >= 0 && wp->w_buffer->b_virt_text_inline > 0) {
-    marktree_itr_get(wp->w_buffer->b_marktree, cts->cts_row, 0, cts->cts_iter);
-    MTKey mark = marktree_itr_current(cts->cts_iter);
-    if (mark.pos.row == cts->cts_row) {
-      cts->cts_has_virt_text = true;
+  cts->win = wp;
+  cts->line = line;
+  cts->max_head_vcol = 0;
+  cts->cur_text_width_left = 0;
+  cts->cur_text_width_right = 0;
+  cts->virt_row = -1;
+  cts->indent_width = INT_MIN;
+  cts->special_tab = !wp->w_p_list || wp->w_p_lcs_chars.tab1;
```

Comment:
`special_tab` doesn't seem an accurate name for this. `use_tabstop` seems better.